// kotrans/Server/kotrans.server.js
/**
 * server.connection.js
 * 
 * Server connection using node.js. To utilize this file, user must invoke the 
 * require() nodejs method.
 *
 * require('path/to/server.connection.js');
 * 
 * @author  Sam Ko
 */

'use strict';

var kotrans = kotrans || {};

kotrans.server = (function () {
    var fs = require('fs');
    var os = require('os');
    var cluster = require('cluster');
    var exec = require('child_process').exec;
    var child_process = require('child_process');
    var http = require( 'http' );
    var path = require('path');
    var BinaryServer = require('binaryjs').BinaryServer;
    var async = require('async');
    //done signifies all files were transfered
    var Client2ServerFlag = {
        send : 'send',
        transferComplete: 'transferComplete',
        setup : 'setup',
        stream_id : 'stream_id',
        main_id : 'main_id'
    }

    //sent signifies that the file chunk was sent.
    var Server2ClientFlag = {
        sent: 'sent',
        updateClient: 'updateClient',
        commandComplete: 'commandComplete',
        error: 'error',
        setup : 'setup'
    }

	var socketServer;
    var processors;
    var worker;

    var uploadedBytes;
    var fileSize;

    var file;
    var percentComplete;

    var allowedDirectory;

    var mainClientList;
    var streamList;
    var child_list;

    function createServer(options, callback) {
        mainClientList = [];
        streamList = [];
        child_list = [];

        var options = options || {};
        var server = options.server;
        var route = options.route;
        //code for multiple threads means faster concatenation
        processors = os.cpus().length;
        allowedDirectory = options.directory || __dirname + '/../../../uploads';

        socketServer = new BinaryServer({ server: server, path : route });

        socketServer.on('connection', onSocketConnection);

        socketServer.on('error', function(err) {
            console.log(err);
        })
        if(callback instanceof Function) {
            callback();
        }

        
    }

    //wait for new user connection
    var execQueue;
    var initialized = false;
    function onSocketConnection(client) {
        console.log("Client connected with ID " + client.id);

        //work on the incoming stream from browsers
		client.on('stream', function (stream, meta) {
			if (meta.cmd === Client2ServerFlag.send) {
                file = fs.createWriteStream(path.join(allowedDirectory, meta.chunkName));
                uploadedBytes += meta.chunkSize;
                file.on('error', function(err) {
                    console.log(err);
                    process.exit();
                });
				stream.pipe(file);    
			} else if(meta.cmd === Client2ServerFlag.transferComplete) {
                //entire file has finished transferring, combining pieces together
                child_list[meta.main_id].send({
                    cmd : 'finish'
                });
            } else if(meta.cmd === Client2ServerFlag.setup) {
                //only master client will be able to enter this.
                var child;
                uploadedBytes = 0;
                fileSize = meta.fileSize;
                if(child_list[client.id]) {
                    child_list[client.id].kill();
                }
                child = child_process.fork('./worker', {
                    cwd : __dirname
                });

                child.on('message', function(message) {
                    if(message === 'working') {
                        setTimeout(function() {
                            child.send({
                                cmd : 'finish'
                            });
                        }, 50);
                    } else if(message === 'finished') {
                        client.send({}, { cmd : Server2ClientFlag.commandComplete });
                        child_list[client.id].kill();
                    }
                });

                child.send({
                    cmd : 'setup'
                });

                child.send({
                    cmd : 'directory',
                    directory : allowedDirectory
                });

                child_list[client.id] = child;

            } else if(meta.cmd === Client2ServerFlag.main_id) {
                //let the server know which streams comes from which client.
                mainClientList[client.id] = {
                    streams : []
                }

                client.send({}, {
                    cmd : Client2ServerFlag.main_id,
                    main_id : client.id
                });
            } else if(meta.cmd === Client2ServerFlag.stream_id) {
                mainClientList[meta.main_id].streams.push(client.id);
                console.log(mainClientList)
                client.main_id = meta.main_id;
            }

            // Sends data back to the client with a percentage complete with file name
			stream.on('data', function (data) {
                
			});

            // Send a message to the client that the fileChunk was successfully transferred.
			stream.on('end', function () {
                if(meta.cmd === Client2ServerFlag.send) {
                    child_list[client.main_id].send({
                        cmd : 'file',
                        name : meta.chunkName
                    });
                    client.send({}, { chunkName: meta.chunkName, cmd: Server2ClientFlag.sent });
                }
			});
        });
    }

    module.exports = {
        createServer: function(options, callback) {
            createServer(options, callback);
        }
    }
})();
